ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+
+Furthemore, all of the new behaviour is performed within the body of the invoked subroutine entirely by inspecting the values of the arguments that were passed. The subroutine is not made aware of how those values came to be passed in - whether from literal name-value syntax, a hash or array variable expansion, or any other expression yielding such a list of argument name and value pairs.
+
+```perl
+sub make_colour ( :$red, :$green, :$blue ) { ... }
+
+make_colour( red => 1.0, blue => 0.5, green => 0.2 );
+# The body of the function will be invoked with
+#   $red   = 1.0
+#   $green = 0.2
+#   $blue  = 0.5
+```
+
+As with positional parameters, a named parameter without a defaulting expression is mandatory, and an error will be raised as an exception if the caller fails to pass a corresponding value. A defaulting expression may be specified using any of the operators available to positional parameters - `=`, `//=` or `||=`.
+
+```perl
+sub make_colour ( :$red = 0, :$green = 0, :$blue = 0 ) { ... }
+
+make_colour( red => 1.0, blue => 0.5 );
+# The body of the function will be invoked with
+#   $red   = 1.0
+#   $green = 0
+#   $blue  = 0.5
+```
+
+A subroutine is permitted to use a combination of *mandatory* positional and named parameters in its definition, provided that all named parameters appear after all the positional ones. Any named parameters may be optional; there are no ordering constraints here.
+
+If a subroutine uses named parameters then it may optionally use a slurpy hash argument as its final element. In this case, the hash will receive all *other* name-value pairs passed by the caller, apart from those consumed by named parameters. If the subroutine does not use a slurpy argument, then it will be an error raised as an exception for there to be any remaining name-value pairs after processing.
+
+While all of the above has been specified in terms of subroutines, every point should also apply equally to the methods provided by the `class` feature, after first processing the implied invocant `$self` parameter.
+
+## Backwards Compatibility
+
+At the site of a subroutine's definition, this specification only uses new syntax in the form of the leading colon prefix on parameter names. Such syntax was formerly invalid in previous versions of perl. Thus there is no danger of previously-valid code being misinterpreted.
+
+All of the behaviour provided by this new syntax is compatible with and analogous to any existing code that could have been written prior, perhaps by direct assignment into a hash. There are no visible differences in the external interface to a subroutine using such syntax, and it remains callable in exactly the same way. Functions provided by modules could be upgraded to use the new syntax without any impact on existing code that invokes them.
```

Comment:
This assumes named parameters use hash initializer semantics:

 - Duplicate keys are allowed in calls.
 - The last instance of a particular key "wins" (but the preceding arguments are still evaluated).

Reply:
Mmm. This is fair; there could be other equally-valid implementations. The implementation could warn about duplicate keys, pointing out that earlier values were overwritten by later ones.

It could even be left as an extension idea to e.g. define a ":collect" attribute that makes an array of all the values given to a particular arg:

```perl
sub f ( :@x :collect ) { say "Value $_" for @x; }

f( x => 123, x => 456, x => 789 );
```

It seems fairly niche interest at that point.

Of course, nothing stops a function from simply taking a `'@rest` array and unpacking it themselves; perhaps in a multivalue foreach loop. The aim of this PPC is to try to find a balance of making the common cases easy, without getting in the way of the weird ones.

Reply:
The first thing I saw with `sub f(:@x :collect) {...}` is that I can't quickly tell that someone didn't make a typo meaning `sub f(:@x, :$collect) {...}`:

```perl
sub f(:@x, :$collect) { 
     say "x: $_" for @x;

     # 400+ lines of code

     $payment_processor->get_money(@x) if $collect;
}

f( x => 123, x => 456, x => 789, collect => true );
```

Reply:
Test case suggestion:
```
sub foo (:$x, @x);
foo (x => 1, x => q (slurpy));
```